system/include/SSE/emmintrin.h

```diff
@@ -0,0 +1,1616 @@
+                                   (imm) & 0x3, ((imm) & 0xc) >> 2, \
+                                   ((imm) & 0x30) >> 4, ((imm) & 0xc0) >> 6, \
+                                   4, 5, 6, 7); })
+
+#define _mm_shufflehi_epi16(a, imm) __extension__ ({ \
+  (__m128i)__builtin_shufflevector((__i16x8)(__m128i)(a), \
+                                   (__i16x8)_mm_set1_epi16(0), \
+                                   0, 1, 2, 3, \
+                                   4 + (((imm) & 0x03) >> 0), \
+                                   4 + (((imm) & 0x0c) >> 2), \
+                                   4 + (((imm) & 0x30) >> 4), \
+                                   4 + (((imm) & 0xc0) >> 6)); })
+
+static __inline__ __m128i __attribute__((__always_inline__, __nodebug__))
+_mm_unpackhi_epi8(__m128i __a, __m128i __b)
+{
+  return (__m128i)__builtin_shufflevector((__i8x16)__a, (__i8x16)__b, 8, 16+8, 9, 16+9, 10, 16+10, 11, 16+11, 12, 16+12, 13, 16+13, 14, 16+14, 15, 16+15);
+}
+
+static __inline__ __m128i __attribute__((__always_inline__, __nodebug__))
+_mm_unpackhi_epi16(__m128i __a, __m128i __b)
+{
+  return (__m128i)__builtin_shufflevector((__i16x8)__a, (__i16x8)__b, 4, 8+4, 5, 8+5, 6, 8+6, 7, 8+7);
+}
+
+static __inline__ __m128i __attribute__((__always_inline__, __nodebug__))
+_mm_unpackhi_epi32(__m128i __a, __m128i __b)
+{
+  return (__m128i)__builtin_shufflevector((__i32x4)__a, (__i32x4)__b, 2, 4+2, 3, 4+3);
+}
+
+static __inline__ __m128i __attribute__((__always_inline__, __nodebug__))
+_mm_unpackhi_epi64(__m128i __a, __m128i __b)
+{
+  return (__m128i)__builtin_shufflevector((__i32x4)__a, (__i32x4)__b, 2, 3, 4+2, 4+3);
+}
+
+static __inline__ __m128i __attribute__((__always_inline__, __nodebug__))
+_mm_unpacklo_epi8(__m128i __a, __m128i __b)
+{
+  return (__m128i)__builtin_shufflevector((__i8x16)__a, (__i8x16)__b, 0, 16+0, 1, 16+1, 2, 16+2, 3, 16+3, 4, 16+4, 5, 16+5, 6, 16+6, 7, 16+7);
+}
+
+static __inline__ __m128i __attribute__((__always_inline__, __nodebug__))
+_mm_unpacklo_epi16(__m128i __a, __m128i __b)
+{
+  return (__m128i)__builtin_shufflevector((__i16x8)__a, (__i16x8)__b, 0, 8+0, 1, 8+1, 2, 8+2, 3, 8+3);
+}
+
+static __inline__ __m128i __attribute__((__always_inline__, __nodebug__))
+_mm_unpacklo_epi32(__m128i __a, __m128i __b)
+{
+  return (__m128i)__builtin_shufflevector((__i32x4)__a, (__i32x4)__b, 0, 4+0, 1, 4+1);
+}
+
+static __inline__ __m128i __attribute__((__always_inline__, __nodebug__))
+_mm_unpacklo_epi64(__m128i __a, __m128i __b)
+{
+  return (__m128i)__builtin_shufflevector((__i32x4)__a, (__i32x4)__b, 0, 1, 4+0, 4+1);
+}
+
+static __inline__ __m128i __attribute__((__always_inline__, __nodebug__))
+_mm_move_epi64(__m128i __a)
+{
+  return __builtin_shufflevector((__i32x4)__a, (__i32x4)wasm_i32x4_const(0, 0, 0, 0), 0, 1, 4+0, 4+1);
+}
+
+static __inline__ __m128d __attribute__((__always_inline__, __nodebug__))
+_mm_unpackhi_pd(__m128d __a, __m128d __b)
+{
+  return (__m128d)wasm_v64x2_shuffle((v128_t)__a, (v128_t)__b, 1, 3);
```

Comment:
Why are some of these using `__builtin_shufflevector` and others using the shuffle intrinsics? Also, where did we get with the investigation of why the intrinsics are performing worse? Is there a particular function I should look at where I would be able to see a difference?

Reply:
Migrated to use the wasm shuffles.

I think looking at the microbenchmarks will be a good way to identify particularly interesting functions to look at. Though for SSE2 the microbenchmark is not yet complete.

Reply:
Oh, one thing I noticed is that the wasm shuffles macros is a bit unsafe. E.g.
```
#define wasm_v64x2_shuffle(__a, __b, __c0, __c1)                               \
  ((v128_t)(                                                                   \
      __builtin_shufflevector((__u64x2)(__a), (__u64x2)(__b), __c0, __c1)))
```
to
```
#define wasm_v64x2_shuffle(__a, __b, __c0, __c1)                               \
  ((v128_t)(                                                                   \
      __builtin_shufflevector((__u64x2)(__a), (__u64x2)(__b), (__c0), (__c1))))
```
Added extra guard parens to the caller side in emmintrin.h to work around.

Reply:
Are they unsafe even though those unparenthesized arguments have to be integer literals? I'll add parens to the upstream header to err on the side of caution either way.

Reply:
Ah nevermind, I see the issue. Fixing upstream.